File: scripts/eval/phase3_future_influence.py
"""Phase 3.4 Future Influence Experiment."""

import argparse
import json
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from longmem_dmr_smoke import eval_cargo_profile, repo_root


def run_phase3_future_influence(output_path: Path, tag: str) -> dict:
    cargo_profile = eval_cargo_profile()
    cmd = ["cargo", "run"]
    if cargo_profile == "release":
        cmd.append("--release")
    cmd.extend(
        [
            "-p",
            "synapse-eval",
            "--bin",
            "kr-phase3-future-influence",
            "--",
            "--json",
            str(output_path),
            "--tag",
            tag,
        ]
    )
    result = subprocess.run(cmd, cwd=repo_root(), text=True, capture_output=True)
    if result.returncode != 0:
        raise RuntimeError(
            "kr-phase3-future-influence failed\nSTDOUT:\n{}\nSTDERR:\n{}".format(
                result.stdout, result.stderr
            )
        )
    print(result.stdout, end="")
    if result.stderr:
        print(result.stderr, end="", file=sys.stderr)
    report = json.loads(output_path.read_text(encoding="utf-8"))
    report["cargo_profile"] = cargo_profile
    return report


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Run Phase 3.4 future influence evaluation."
    )
    parser.add_argument(
        "--output-path",
        type=Path,
        default=repo_root() / "crates/eval/reports/phase3-future-influence.json",
    )
    parser.add_argument("--tag", default="phase3-future-influence")
    args = parser.parse_args()

    args.output_path.parent.mkdir(parents=True, exist_ok=True)
    report = run_phase3_future_influence(args.output_path, args.tag)
    print(
        "summary: scenarios={} helpful={} neutral={} rejected={} gain={:.4f} decision={:.4f} failure_reduction={:.4f} usefulness={:.4f} safety={:.4f} pass={}".format(
            report["scenarios"],
            report["results"]["helpful_lessons"],
            report["results"]["neutral_lessons"],
            report["results"]["rejected_influence"],
            report["metrics"]["influence_gain_score"],
            report["metrics"]["decision_improvement_score"],
            report["metrics"]["failure_reduction_score"],
            report["metrics"]["lesson_usefulness_score"],
            report["metrics"]["no_write_safety"],
            report["pass"],
        )
    )
    print(f"Saved to {args.output_path}")
    if report["pass"] and report["mode"] == "evaluation_only":
        print("Phase 3.4 Future Influence Experiment Completed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/eval/src/bin/phase3_future_influence.rs
use anyhow::{Context, Result};
use clap::Parser;
use std::path::PathBuf;
use synapse_eval::Phase3FutureInfluenceEvaluator;

#[derive(Parser)]
#[command(
    name = "kr-phase3-future-influence",
    about = "Synapse Phase 3.4 future influence evaluation"
)]
struct Cli {
    /// Write the Phase3FutureInfluenceReport JSON to this path.
    #[arg(long)]
    json: PathBuf,
    /// Tag to embed in the report.
    #[arg(long, default_value = "phase3-future-influence")]
    tag: String,
}

fn main() -> Result<()> {
    let cli = Cli::parse();
    let report = Phase3FutureInfluenceEvaluator::evaluate(cli.tag)?;

    println!("=== Synapse Phase 3.4 future influence experiment ===");
    println!("tag:                              {}", report.tag);
    println!("phase:                            {}", report.phase);
    println!("mode:                             {}", report.mode);
    println!(
        "evaluation_version:               {}",
        report.evaluation_version
    );
    println!(
        "baseline_version:                 {}",
        report.baseline_version
    );
    println!("scenarios:                        {}", report.scenarios);
    println!(
        "promoted_lesson_source_count:     {}",
        report.promoted_lesson_source_count
    );
    println!(
        "helpful_lessons:                  {}",
        report.results.helpful_lessons
    );
    println!(
        "neutral_lessons:                  {}",
        report.results.neutral_lessons
    );
    println!(
        "rejected_influence:               {}",
        report.results.rejected_influence
    );
    println!(
        "influence_gain_score:             {:.4}",
        report.metrics.influence_gain_score
    );
    println!(
        "decision_improvement_score:       {:.4}",
        report.metrics.decision_improvement_score
    );
    println!(
        "failure_reduction_score:          {:.4}",
        report.metrics.failure_reduction_score
    );
    println!(
        "lesson_usefulness_score:          {:.4}",
        report.metrics.lesson_usefulness_score
    );
    println!(
        "no_write_safety:                  {:.4}",
        report.metrics.no_write_safety
    );
    println!(
        "memory_written:                   {}",
        report.safety.memory_written
    );
    println!(
        "playbook_created:                 {}",
        report.safety.playbook_created
    );
    println!(
        "future_influence_changed:         {}",
        report.safety.future_influence_changed
    );
    println!("pass:                             {}", report.pass);

    if let Some(parent) = cli.json.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(&cli.json, serde_json::to_string_pretty(&report)?)
        .with_context(|| format!("writing JSON to {}", cli.json.display()))?;
    eprintln!(
        "wrote Phase 3.4 future influence report to {}",
        cli.json.display()
    );
    Ok(())
}
